### PR TITLE
simplify IO rate limiter algorithm

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

I read through the code of https://github.com/tikv/async-speed-limit, and realize the algorithm for `IORateLimiter` can be greatly simplified with one signed integer counter instead of two unsigned ones. One catch is race between request enqueue and budget refill, but we can always reason the ordering by treating the actual atomic refill as epoch boundary.

Also simplify the handling of excessively long waits. Now the consumption counter will only be monotonically updated.

### Check List 

Tests
- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/file_system/src/rate_limiter.rs [hunk 1]
```diff
@@ -325,16 +225,88 @@ impl PriorityBasedIORateLimiter {
...
+            wait = MAX_WAIT_DURATION_PER_REQUEST;
+            amount = std::cmp::max(
+                (MAX_WAIT_DURATION_PER_REQUEST.as_secs_f32() * amount as f32 / wait.as_secs_f32())
+                    as usize,
+                1,
+            );
+            // Subtracting redundant bytes requested before.
+            let redundant = remains as i64 - amount as i64;
+            debug_assert!(redundant > 0);
+            self.bytes_available[priority_idx].fetch_add(redundant, Ordering::Relaxed);
```
Comment (Connor1996) [inline components/file_system/src/rate_limiter.rs:307]: seems equal to `self.bytes_available[priority_idx].store(amount)`?

File: components/file_system/src/rate_limiter.rs [hunk 2]
```diff
@@ -345,68 +317,49 @@ impl PriorityBasedIORateLimiter {
...
+        if budgets == 0 {
             // It's possible that rate limit is toggled off in the meantime.
             return;
         }
         debug_assert!(now >= locked.next_refill_time);
-        let skipped_epochs =
-            (now - locked.next_refill_time).as_secs_f32() / DEFAULT_REFILL_PERIOD.as_secs_f32();
+        let elapsed_epochs = (now - locked.next_refill_time).as_secs_f32()
+            / DEFAULT_REFILL_PERIOD.as_secs_f32()
+            + 1.0;
```
Comment (Connor1996) [inline components/file_system/src/rate_limiter.rs:328]: why plus 1.0

File: components/file_system/src/rate_limiter.rs [hunk 3]
```diff
@@ -345,68 +317,49 @@ impl PriorityBasedIORateLimiter {
...
-                RATE_LIMITER_MAX_BYTES_PER_SEC
-                    .low
-                    .set((remaining_budgets * DEFAULT_REFILLS_PER_SEC) as i64);
-            }
-            self.bytes_per_epoch[p - 1].store(remaining_budgets, Ordering::Relaxed);
+            let max_refill = (budgets as f32 * elapsed_epochs) as i64;
+            let unused = self.bytes_available[p].load(Ordering::Relaxed);
+            let til_full = budgets as i64 - unused;
+            self.bytes_available[p]
+                .fetch_add(std::cmp::min(max_refill, til_full), Ordering::Relaxed);
```
Comment (Connor1996) [inline components/file_system/src/rate_limiter.rs:360]: use `fetch_update`

File: components/file_system/src/rate_limiter.rs (hunk 2 above)
Comment (tabokie) [reply components/file_system/src/rate_limiter.rs:328]: `elapsed == now - last_refill_time == now - next_refill_time + one_epoch`

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by tabokie** [comment]
Closing this temporarily.

**Pull request closed without merging**